### DISCUSSION TITLE
Doubts regarding Multioutput Gaussian Process in Gpytorch and GPU Memory Usage

### DISCUSSION THREAD

**Discussion by saavnkr** [cornellius-gp/gpytorch#2365] [Q&A]
Hi all,

I'm new to Gpytorch, so please forgive me if my questions seem too basic. I'm currently using the Multioutput Gaussian Process (Multitask GP) for handling inputs and outputs with shapes `(N, D)`, where` N` represents the number of samples (in this case, images) and `D` represents the corresponding flattened pixels.

I have two specific doubts:

1. In the prediction phase, I have noticed that the columns in the var_pred matrix `(shape: (N_t, D))` have the same values. In other words, the model is not learning the covariance in the` N` direction. As a result, when I plot the variance, it simply appears as noise.

2. When I use `observed_pred = likelihood(model(x_t))`, it consumes a significantly large amount of GPU memory (90+ GB). However, when I use only `likelihood(x_t)`, it consumes much less GPU memory. Here, `x_t` represents the test data with shape (N, D).

Could you please help me understand these issues and provide any possible solutions?

**Comment by gpleiss** [comment]
> In the prediction phase, I have noticed that the columns in the var_pred matrix (shape: (N_t, D)) have the same values. In other words, the model is not learning the covariance in the N direction. As a result, when I plot the variance, it simply appears as noise.

This may very well be the case, depending on the model/data that you are trying to represent. Just like any machine learning model, a GP can fall into bad local minima if your data are complex or not preprocessed in an appropriate way. Many GP kernels can also struggle with high dimensional data.

I would z-score your data, and try scaling down to a small problem (i.e. set D=10), and see if the issue persists.

> When I use observed_pred = likelihood(model(x_t)), it consumes a significantly large amount of GPU memory (90+ GB). However, when I use only likelihood(x_t), it consumes much less GPU memory. Here, x_t represents the test data with shape (N, D).

You need to run `likelihood(model(x_t))` - this produces the predictive posterior. `likelihood(x_t)` does not do the thing that you expect to do.

There could be many reasons why this is taking lots of GPU memory (GPs are notoriously memory intensive), and there are modifications that you can make to the model to reduce this. If you want any feedback on this, you'll need to give us more information (i.e. what is N, D, what is the data that you are using) as well as a runable code example.
